File: src/cpp/multiprocessing/spawn_process.cpp
#include <iostream>
#include <sys/wait.h>
#include <thread>
#include <unistd.h> // getpid() definition

void foo() {
  std::cout << "Worker function." << std::endl;
  std::cout << "Worker function id: " << getpid() << std::endl;
  std::cout << "Worker thread id: " << std::this_thread::get_id() << std::endl;
}

int main() {
  std::cout << "Main function id: " << getpid() << std::endl;
  std::cout << "Main thread id: " << std::this_thread::get_id() << std::endl;

  // create a new process with fork
  auto pid = fork();
  if (pid == 0) {
    // child process
    foo();
  } else {
    // parent process
    wait(NULL);
  }

  return 0;
}

File: src/python/multiprocessing/spawn_process.py
import os
import multiprocessing
import threading


def foo():
    print("Worker function")
    print(f"Worker function process id: {os.getpid()}")
    print(f"Worker thread id: {threading.currentThread().ident}")
    print(f"Worker thread id: {threading.currentThread().getName()}")


if __name__ == "__main__":

    print(f"Main function process id: {os.getpid()}")
    print(f"Main thread id: {threading.currentThread().ident}")
    print(f"Main thread id: {threading.currentThread().getName()}")

    process = multiprocessing.Process(target=foo)
    process.start()
    process.join()
